### PR TITLE
Update for a docker driver

### PR DESCRIPTION
**Pull request by clintkitson**
This commit enables a fully functioning volume driver that is
able to run from VMs in VirtualBox and perform introspection
necessary to issue commands to underlying VB SOAP API
to perform storage orchestration for volume attachemnt.

### CODE REVIEW

**Comment by md5** [comment]
This looks great! I'm glad you were able to build on my work for this. :+1: 

I haven't had a chance to review this in detail yet, but the first thing that jumps out are the changes to `vboxwebsrv/vboxwebsrv.go`. Since that file was generated by [`gowsdl`](https://github.com/hooklift/gowsdl), it may be better to make the changes there and regenerate `vboxwebsrv.go`.

I'll find some time to review this more fully in the coming week.

**Comment by clintkitson** [comment]
Btw.. just noticed this is a v4 thing only.  The API changed in v5, which I have adjusted for.  But I am running into an issue locking the instances while running.  I developed the functionality against VirtualBox 4.3.28.

```
dicey1:siomachine clintonkitson$ VBoxManage storageattach ad92d16a-ef7b-4b4a-abb2-31903dde203f --storagectl "SATA Controller" --port 1 --device 0 --type hdd --medium 4c2a05ee-5a50-4665-9633-2b4530f3cc35
VBoxManage: error: The machine is not mutable (state is Running)
VBoxManage: error: Details: code VBOX_E_INVALID_VM_STATE (0x80bb0002), component SessionMachine, interface IMachine, callee nsISupports
```

https://www.virtualbox.org/ticket/14936#comment:1

**Comment by md5** [comment]
I think in general a vboxwebsrv client will probably have to deal with
multiple versions of the API. I haven't thought much about how to do that
yet, but it would be cool if there were a stable API call to query the
version. I haven't yet checked to see if that exists, though querying the
WSDL could be an option.

**Comment by clintkitson** [comment]
I think one of the problems may turn into bloat though.  If each SOAP API has its own WSDL with minor changes, then that turns into a ton of extra code as versions change.  It might be worth the effort to identify the differences and manually add new methods etc all consolidated under the same vboxwebsrv.go file.

**Comment by md5** [comment]
I suppose that depends on whether Virtualbox is using semantic versioning for its API or not. I would hope it would be possible to have only V4, V5, etc.

**Comment by md5** [comment]
It looks like `IVirtualBoxgetVersion` is the operation to call, but you need to call `logon` first to get the object to call it on. As long as the `logon` operation stays consistent, this should be doable, but it seems like it could be brittle. I also took a quick look at the WSDL and it doesn't provide any indication of which version of the API it supports...

**Comment by md5** [comment]
I've opened https://github.com/hooklift/gowsdl/pull/50 to propose your `err` handling changes for the generator tool used to generate `vboxwebsrv.go`.

As for the `log.Println` changes, I'd rather not bring those in. What would you thing about having the code that uses `go-virtualboxclient` call `log.SetOutput` to control where the logging goes instead?

File: vboxwebsrv/vboxwebsrv.go [hunk 1]
```diff
@@ -15800,7 +15800,7 @@ type IAdditionsFacility struct {
 }
 
 type IMediumAttachment struct {
-	XMLName xml.Name `xml:"http://www.virtualbox.org/ IMediumAttachment"`
+	// XMLName xml.Name `xml:"http://www.virtualbox.org/ IMediumAttachment"`
```
Comment (md5) [inline vboxwebsrv/vboxwebsrv.go:15803]: Do you recall why you commented this out?

File: vboxwebsrv/vboxwebsrv.go [hunk 2]
```diff
@@ -20347,6 +20347,8 @@ func (service *VboxPortType) IMachinegetMedium(request *IMachinegetMedium) (*IMa
 //   - InvalidObjectFault
 //   - RuntimeFault
 
+type Returnval *IMediumAttachment
```
Comment (md5) [inline vboxwebsrv/vboxwebsrv.go:20350]: Could you explain this? It seems unused.

**Comment by md5** [comment]
@clintonskitson FYI, https://github.com/hooklift/gowsdl/pull/50 has been merged. Good catch on the `err` handling!

File: vboxwebsrv/vboxwebsrv.go (hunk 1 above)
Comment (clintkitson) [inline vboxwebsrv/vboxwebsrv.go:15803]: Yup agree with your comment above for loglevel.  This is necessary due to an error that was being returned.  It was throwing a returnval error which seemed aligned to not parsing the information correctly.

File: vboxwebsrv/vboxwebsrv.go (hunk 2 above)
Comment (clintkitson) [inline vboxwebsrv/vboxwebsrv.go:20350]: Remnant of troubleshooting.

**Comment by clintkitson** [comment]
Verry cool.  I have the 5.0.10 WSDL all dialed in and working for the use case.  I can re-run it and make the minor tweak re commenting out the XML def as referenced above.  Wondering if you would rather have that as the PR for now and then deal with backward/forward versioning in the future?  Maybe just leave them as separate branches for now?

**Comment by md5** [comment]
I'm happy to have this PR just targeting the V4 binding. Opening a branch for V5 compatibility also seems like a reasonable approach until I have the time to work out a better approach. I've created a [`v5-api`](https://github.com/appropriate/go-virtualboxclient/tree/v5-api) branch if you want to send a PR that way.

Regarding the `XMLName` removal, I'm hesitant to accept a patch that removes it. If you could provide some more information on how it was failing, perhaps including the returned `SOAPBody`, that would likely help.

**Comment by md5** [comment]
Actually, perhaps it would be better to have `master` be V5 and have a `v4-api` branch...

What do you think?

**Comment by clintkitson** [comment]
Yeah, I think keeping master at the latest would be the ideal approach.

**Comment by md5** [comment]
OK. I've killed `v5-api` and created [`v4-api`](https://github.com/appropriate/go-virtualboxclient/tree/v4-api).

**Comment by md5** [comment]
BTW, were there any adjustments needed to the code under `virtualboxclient` based on the V4 v. V5 api changes?

**Comment by clintkitson** [comment]
Not for the direct bindings.  But there was a helper function for creating
storage that did need to be updated.

**Comment by clintkitson** [comment]
Opened https://github.com/appropriate/go-virtualboxclient/pull/2 and https://github.com/appropriate/go-virtualboxclient/pull/3

**Pull request closed without merging**